File: concrete/src/main/java/pcb/uwu/amount/derived/termodynamics/Watts.java
package pcb.uwu.amount.derived.termodynamics;

import pcb.uwu.amount.base.Amperes;
import pcb.uwu.amount.base.ElectricCurrent;
import pcb.uwu.amount.base.Time;
import pcb.uwu.amount.derived.electromagnetism.ElectricPotential;
import pcb.uwu.amount.derived.electromagnetism.Volts;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.derived.termodynamics.PowerUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.unit.base.AmpereUnit.AMPERE;
import static pcb.uwu.unit.base.SecondUnit.SECOND;
import static pcb.uwu.unit.derived.electromagnetism.VoltUnit.VOLT;
import static pcb.uwu.unit.derived.termodynamics.WattUnit.WATT;

public class Watts extends Power {

	// region constructors

	public Watts(Number value) {
		super(value, WATT);
	}

	public Watts(Number value, Magnitude magnitude) {
		super(value, magnitude, WATT);
	}

	public Watts(String value) {
		super(value, WATT);
	}

	public Watts(String value, Magnitude magnitude) {
		super(value, magnitude, WATT);
	}

	public Watts(BigDecimal value) {
		super(value, WATT);
	}

	public Watts(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, WATT);
	}

	public Watts(BigDecimalAmount amount) {
		super(amount, WATT);
	}

	public Watts(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, WATT);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Watts plus(UnitAmount<PowerUnit> other, MathContext mathContext) {
		return new Watts(plusAmount(other, mathContext));
	}

	@Override
	public Watts minus(UnitAmount<PowerUnit> other, MathContext mathContext) {
		return new Watts(minusAmount(other, mathContext));
	}

	@Override
	public Watts multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Watts(multipliedByScalar(other, mathContext));
	}

	@Override
	public Watts dividedBy(BigDecimal other, MathContext mathContext) {
		return new Watts(dividedByScalar(other, mathContext));
	}

	// endregion

	// region composition

	public Joules multipliedBy(Time time, MathContext mathContext) {
		return new Joules(getAmount().multipliedBy(time.getAmountIn(SECOND), mathContext));
	}

	public Volts dividedBy(ElectricCurrent electricCurrent, MathContext mathContext) {
		return new Volts(getAmount().dividedBy(electricCurrent.getAmountIn(AMPERE), mathContext));
	}

	public Amperes dividedBy(ElectricPotential electricPotential, MathContext mathContext) {
		return new Amperes(getAmount().dividedBy(electricPotential.getAmountIn(VOLT), mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/derived/thermodynamics/Watts.kt
package pcb.uwu.amount.derived.thermodynamics

import pcb.uwu.amount.base.Amperes
import pcb.uwu.amount.base.ElectricCurrent
import pcb.uwu.amount.base.Time
import pcb.uwu.amount.derived.electromagnetism.ElectricPotential
import pcb.uwu.amount.derived.electromagnetism.Volts
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.AmpereUnit
import pcb.uwu.unit.base.SecondUnit
import pcb.uwu.unit.derived.electromagnetism.VoltUnit
import pcb.uwu.unit.derived.termodynamics.PowerUnit
import pcb.uwu.unit.derived.termodynamics.WattUnit

class Watts : Power
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, WattUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, WattUnit)

    // region UnitAmount

    override fun plus(power: UnitAmount<PowerUnit>) =
        Watts(this.amount + (power to WattUnit).amount)

    override fun minus(power: UnitAmount<PowerUnit>) =
        Watts(this.amount - (power to WattUnit).amount)

    override fun times(scalar: Number) =
        Watts(this.amount * scalar)

    override fun div(scalar: Number) =
        Watts(this.amount / scalar)

    // endregion

    // region composition

    override fun times(time: Time) =
        Joules(this.amount * (time to SecondUnit).amount)

    override fun div(electricCurrent: ElectricCurrent) =
        Volts(this.amount / (electricCurrent to AmpereUnit).amount)

    operator fun div(electricPotential: ElectricPotential) =
        Amperes(this.amount / (electricPotential to VoltUnit).amount)

    // endregion
}
